src/realm/table.cpp

```diff
@@ -2210,72 +2180,94 @@ Decimal128 Table::average_decimal(ColKey col_key, size_t* value_count) const
 
 int64_t Table::minimum_int(ColKey col_key, ObjKey* return_ndx) const
 {
+    QueryStateMin<int64_t> st;
     if (is_nullable(col_key)) {
-        return aggregate<act_Min, util::Optional<int64_t>, int64_t>(col_key, 0, nullptr, return_ndx);
+        aggregate<util::Optional<int64_t>>(st, col_key);
     }
-    return aggregate<act_Min, int64_t, int64_t>(col_key, 0, nullptr, return_ndx);
+    else {
+        aggregate<int64_t>(st, col_key);
+    }
+    if (return_ndx) {
+        *return_ndx = st.m_minmax_index;
+    }
+    return st.m_state;
 }
 
 float Table::minimum_float(ColKey col_key, ObjKey* return_ndx) const
 {
-    return aggregate<act_Min, float, float>(col_key, 0.f, nullptr, return_ndx);
+    QueryStateMin<float> st;
+    aggregate<float>(st, col_key);
+    if (return_ndx) {
+        *return_ndx = st.m_minmax_index;
+    }
+    return st.m_state;
 }
 
 double Table::minimum_double(ColKey col_key, ObjKey* return_ndx) const
 {
-    return aggregate<act_Min, double, double>(col_key, 0., nullptr, return_ndx);
+    QueryStateMin<double> st;
+    aggregate<double>(st, col_key);
+    if (return_ndx) {
+        *return_ndx = st.m_minmax_index;
+    }
+    return st.m_state;
 }
 
 Decimal128 Table::minimum_decimal(ColKey col_key, ObjKey* return_ndx) const
 {
-    ArrayDecimal128 leaf(get_alloc());
-    Decimal128 min("+Inf");
-    ObjKey ret_key;
-    auto f = [&min, &ret_key, &leaf, col_key](const Cluster* cluster) {
-        // direct aggregate on the leaf
-        cluster->init_leaf(col_key, &leaf);
-        auto sz = leaf.size();
-        for (size_t i = 0; i < sz; i++) {
-            auto val = leaf.get(i);
-            if (!val.is_null() && val < min) {
-                min = val;
-                ret_key = cluster->get_real_key(i);
-            }
-        }
-        return false;
-    };
-
-    traverse_clusters(f);
+    QueryStateMin<Decimal128> st;
+    aggregate<Decimal128>(st, col_key);
     if (return_ndx) {
-        *return_ndx = ret_key;
+        *return_ndx = st.m_minmax_index;
```

Comment:
I guess the naming should be about keys instead of indexes now.

Reply:
Done